testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/MultiEnvTestEngine.java

```diff
@@ -116,18 +118,53 @@ public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId
                   for (TestDescriptor child : children) {
                     // Note: this also removes the reference to parent from the child
                     discoverResult.removeChild(child);
-                    envRoot.addChild(child);
+
+                    // Must check whether the test class is a multi-env test, because discovery
+                    // returns all test classes.
+                    ClassBasedTestDescriptor classBased = (ClassBasedTestDescriptor) child;
+                    boolean multi = registry().isMultiEnvClass(classBased);
+                    if (multi) {
+                      envRoot.addChild(child);
+                    }
+                  }
+                }
+              });
+
+      // Also execute all other tests via the MultiEnv test engine to get the "thread per
+      // test-class" behavior.
+      registry()
+          .probablyNotMultiEnv()
```

Comment:
Why collect test descriptors in an "extension" registry? Why not reuse `preliminaryResult` here and filter out all multi-env tests?

We already rediscover tests for each env. ID, so we're probably not saving much time by collecting test descriptors during the extension registration process... On the other hand using a discovery result directly would be much more clear... WDYT?

Reply:
The result is sadly not correct that way. We can make it "perfect" in a follow-up.